### ISSUE TITLE
More helpful error messages when pipeline fails to build or compute?

### ISSUE THREAD

**Issue by nvaytet** [scipp/sciline#74]
When a pipeline fails to build or compute, the error message gives you the type it failed with.
Not sure how feasible it is, but it would be even more helpful if it printed the name of the provider it failed in, as the same types can be used by multiple providers.

This would make debugging pipelines a lot faster.

**Comment by SimonHeybrock** [comment]
Use Python 3.10 or higher, so `NewType` is implemented as a class instead of a function. You will then get proper errors.

As Python 3.8 and 3.9 are/will be dropped soon I will close this.

**Comment by jokasimr** [comment]
It seems feasible. We could add both `tp` and the provider that required `tp` to the stack [here](https://github.com/scipp/sciline/blob/2547d90fd85c1b80d6a2521b205d80bd97fecd65/src/sciline/pipeline.py#L549) and then make the provider available when the error is raised, either by passing it as an extra argument to `_get_provider` or by catching the exceptions raised by `_get_provider` and editing the error messages to add the name of the provider.

**Comment by SimonHeybrock** [comment]
Conceptually this is not really (or easily) possible: There may be multiple providers that require a certain type. The way the graph is build (non-recursively) we do not necessary know which provider "required" something. We could add tracking via list of dependees. See
https://github.com/scipp/sciline/blob/2547d90fd85c1b80d6a2521b205d80bd97fecd65/src/sciline/pipeline.py#L546-L549

Would it make more sense to generalize other available diagnostic tools? See also my comment https://github.com/scipp/sciline/pull/86#discussion_r1429595333.

Basically, #74 and #86 are just two statements about "invalid" graphs: Graph is disconnected, or graph is incomplete. By adding "piecemeal" solutions we risk making the code hard to maintain. I suggest we take a step back and think about the bigger picture before moving forward with any of these.
